### ISSUE TITLE
500 error when exporting large dataset to Excel

### ISSUE THREAD

**Issue by PieterCooreman** [espocrm/espocrm#2209]
**Describe the bug**
Exporting 50.000 contacts to XSLX results in a 500 error. CSV works fine.

**EspoCRM version**
7.0.9

**Additional context**
PHP crash log: 
[24-Jan-2022 21:08:46 UTC] PHP Fatal error:  Allowed memory size of 268435456 bytes exhausted (tried to allocate 20480 bytes) in D:\inetpub\wwwroot\espocrm\application\Espo\Core\InjectableFactory.php on line 166
[24-Jan-2022 21:15:47 UTC] PHP Fatal error:  Allowed memory size of 536870912 bytes exhausted (tried to allocate 8192 bytes) in D:\inetpub\wwwroot\espocrm\application\Espo\ORM\SthCollection.php on line 138
[24-Jan-2022 21:18:45 UTC] PHP Fatal error:  Allowed memory size of 1073741824 bytes exhausted (tried to allocate 8192 bytes) in D:\inetpub\wwwroot\espocrm\application\Espo\ORM\BaseEntity.php on line 270
[24-Jan-2022 21:21:08 UTC] PHP Fatal error:  Allowed memory size of 1572864000 bytes exhausted (tried to allocate 20480 bytes) in D:\inetpub\wwwroot\espocrm\vendor\phpoffice\phpspreadsheet\src\PhpSpreadsheet\Collection\Cells.php on line 157

As you see, I tried with 256, 512, 1024 and 1600M for PHP memory_limit - PHP 7.4.27

**Comment by yurikuzn** [comment]
Could you help with investigation?

**Comment by PieterCooreman** [comment]
Sure Yuri! I noticed memory goes completely wild while exporting and PHP eventually crashes. 

Let me know how I can help. Thanks!

**Comment by o-data** [comment]
Just reading through the Milestone update; try adding this patch and test it Pieter, perhaps it been solved already?

#2197

**Comment by o-data** [comment]
Just reading through the Milestone update; try adding this patch and test it Pieter, perhaps it been solved already?

#2197

**Comment by yurikuzn** [comment]
First, we need to find where memory usage is not optimized. Whether it's in Espo code or PhpSpreadsheet.

**Comment by PieterCooreman** [comment]
I now set memory limit to 2GB -> results in a never ending hanging page. Max memory usage seems to peak at 1GB tough, while processor is 50%. I'm on a Windows 2019 Server btw...

**Comment by PieterCooreman** [comment]
I uploaded the latest codebase but that does not help... Is there anything else that I can do?

**Comment by analystcmyk** [comment]
As of https://github.com/PHPOffice/PhpSpreadsheet/blob/b05dc318507182ef6fe72717bc0f9c9b5ba4bf65/docs/faq.md this is PhpSpreadsheet behaviour. Try to run PhpSpreadsheet locally with a sample on a similar size dataset.

**Comment by PieterCooreman** [comment]
yep, looks like you found the guilty one... I tested 50.000 contacts export with 2GB of RAM... didn't work... at least... I've killied the IIS process as it took way too much processor away... not sure how to deal with this ... back you guys... Maybe disable export to xlsx as from 5.000 records, and offer csv only?

**Comment by yurikuzn** [comment]
It will be processed by CLI php as of v7.1. So it should not kill the webserver process.

**Comment by chaitanya-uc** [comment]
> Author

I think it means that the script for export will execute in the backend so that the server will not timeout.

**Comment by hturkoz** [comment]
Hi,
in the file : vendor/phpoffice/phpspreadsheet/src/PhpSpreadsheet/Collection

namespace PhpOffice\PhpSpreadsheet\Collection;

ini_set('memory_limit',0);

class Cells{ ...

error 500 occur for me and php error memory allocation exhaust...
i have put : ini_set('memory_limit',0); ... result : no error with +- 30000 record

**Comment by kczereczon** [comment]
Maybe we should use chunks of large select (with limit) and processing it chunk by chunk with limit included?

The idea is included here: https://stackoverflow.com/a/34180941

**Comment by jeamado** [comment]
I had the same problem these days but for only 1900 records. I compared Espo 7.08 against Espo 7.15. In Espo version 7.08 those records were exported to .xlsx directly, without  any problem (the new idle feature was not implementes yet). But, for the same dataset, in version 7.1.5 the process stayed for ever "running". In these tests I discovered that the threshold value for passing form export directly to export using cron is 1.000 records. These tests were done in two instances in the same server Centos, PHP 7.4, and of course witth PhpSpreadsheet. I had not memory faults at all, so I wonder if actually this is, or not ,a bug. Could be great for my users to go back to the previos version, or maybe to have the option to enble/disable this cron/idle feature as administrator.
